File: packages/snow_draw_engine/lib/draw/elements/types/arrow/arrow_core_geometry_adapter.dart
import 'package:snow_draw_arrow_core/snow_draw_arrow_core.dart' as core;

import '../../../types/draw_point.dart';
import '../../../types/draw_rect.dart';
import '../../../types/element_style.dart';
import 'arrow_core_codec.dart';

/// Result of recomputing element geometry from edited arrow points.
typedef ArrowGeometryUpdate = ({
  DrawRect rect,
  List<DrawPoint> normalizedPoints,
});

/// Cached layout payload for 2-point arrow-like elements.
///
/// The rect is in world space and [normalizedPoints] are rect-normalized.
class ArrowTwoPointLayout {
  const ArrowTwoPointLayout({
    required this.rect,
    required this.normalizedPoints,
  });

  final DrawRect rect;
  final List<DrawPoint> normalizedPoints;
}

List<DrawPoint> resolveArrowLocalPoints({
  required DrawRect rect,
  required List<DrawPoint> normalizedPoints,
}) {
  final localPoints = core.denormalizeRectPoints(
    rect: <double>[0, 0, rect.width, rect.height],
    normalizedPoints: encodeArrowCorePoints(normalizedPoints),
  );
  return _decodeWithPressure(points: localPoints, source: normalizedPoints);
}

List<DrawPoint> resolveArrowWorldPoints({
  required DrawRect rect,
  required List<DrawPoint> normalizedPoints,
}) {
  final worldPoints = core.denormalizeRectPoints(
    rect: _toBounds(rect),
    normalizedPoints: encodeArrowCorePoints(normalizedPoints),
  );
  return _decodeWithPressure(points: worldPoints, source: normalizedPoints);
}

List<DrawPoint> normalizeArrowPoints({
  required List<DrawPoint> worldPoints,
  required DrawRect rect,
}) {
  final normalized = core.normalizeRectPoints(
    rect: _toBounds(rect),
    worldPoints: encodeArrowCorePoints(worldPoints),
  );
  return _decodeWithPressure(points: normalized, source: worldPoints);
}

DrawRect calculateArrowPathBoundsViaCore({
  required List<DrawPoint> worldPoints,
  required ArrowType arrowType,
}) {
  final bounds = core.calculateArrowPathBounds(
    points: encodeArrowCorePoints(worldPoints),
    curved: arrowType == ArrowType.curved,
  );
  return DrawRect(
    minX: bounds[0],
    minY: bounds[1],
    maxX: bounds[2],
    maxY: bounds[3],
  );
}

/// Computes rect + normalized points for a 2-point arrow fast path.
ArrowTwoPointLayout computeArrowTwoPointLayout({
  required DrawPoint first,
  required DrawPoint second,
}) {
  final layout = core.computeTwoPointRectNormalizedPoints(
    first: <double>[first.x, first.y],
    second: <double>[second.x, second.y],
  );
  return ArrowTwoPointLayout(
    rect: DrawRect(
      minX: layout.rect[0],
      minY: layout.rect[1],
      maxX: layout.rect[2],
      maxY: layout.rect[3],
    ),
    normalizedPoints: List<DrawPoint>.unmodifiable(<DrawPoint>[
      DrawPoint(
        x: layout.normalizedPoints[0][0],
        y: layout.normalizedPoints[0][1],
        pressure: first.pressure,
      ),
      DrawPoint(
        x: layout.normalizedPoints[1][0],
        y: layout.normalizedPoints[1][1],
        pressure: second.pressure,
      ),
    ]),
  );
}

/// Recomputes rect + normalized points after arrow local points changed.
ArrowGeometryUpdate resolveArrowGeometryUpdate({
  required List<DrawPoint> localPoints,
  required DrawRect oldRect,
  required double rotation,
  required ArrowType arrowType,
}) {
  final layout = core.computeRotatedRectAndLocalPoints(
    localPoints: encodeArrowCorePoints(localPoints),
    oldRect: <double>[oldRect.minX, oldRect.minY, oldRect.maxX, oldRect.maxY],
    rotation: rotation,
    curved: arrowType == ArrowType.curved,
  );
  final rect = DrawRect(
    minX: layout.rect[0],
    minY: layout.rect[1],
    maxX: layout.rect[2],
    maxY: layout.rect[3],
  );
  final nextLocalPoints = _decodeWithPressure(
    points: layout.localPoints,
    source: localPoints,
  );
  return (
    rect: rect,
    normalizedPoints: normalizeArrowPoints(
      worldPoints: nextLocalPoints,
      rect: rect,
    ),
  );
}

List<double> _toBounds(DrawRect rect) => <double>[
  rect.minX,
  rect.minY,
  rect.maxX,
  rect.maxY,
];

List<DrawPoint> _decodeWithPressure({
  required List<core.Point> points,
  required List<DrawPoint> source,
}) {
  if (source.isEmpty) {
    return decodeArrowCorePoints(points);
  }
  return List<DrawPoint>.generate(points.length, (index) {
    final sourcePoint =
        source[index < source.length ? index : source.length - 1];
    final point = points[index];
    return DrawPoint(x: point[0], y: point[1], pressure: sourcePoint.pressure);
  }, growable: false);
}

File: packages/snow_draw_rust_engine/src/draw/elements/types/arrow/arrow_core_geometry_adapter.rs
#![allow(dead_code)]

pub use crate::draw::elements::types::arrow::arrow_two_point_layout::{
    compute_arrow_two_point_layout, ArrowTwoPointLayout,
};

use crate::draw::types::draw_point::DrawPoint;
use crate::draw::types::draw_rect::DrawRect;
use crate::draw::types::element_style::ArrowType;

use crate::draw::elements::types::connector::connector_geometry::{
    calculate_connector_rect, normalize_connector_points, resolve_connector_geometry_update,
    resolve_connector_local_points, resolve_connector_world_points, ConnectorGeometryUpdate,
};

/// Resolves arrow local-space points from normalized element-local points.
pub fn resolve_arrow_local_points(
    rect: DrawRect,
    normalized_points: &[DrawPoint],
) -> Vec<DrawPoint> {
    resolve_connector_local_points(rect, normalized_points)
}

/// Resolves arrow world-space points from normalized element-local points.
pub fn resolve_arrow_world_points(
    rect: DrawRect,
    normalized_points: &[DrawPoint],
) -> Vec<DrawPoint> {
    resolve_connector_world_points(rect, normalized_points)
}

/// Normalizes arrow world-space points into element-local space.
pub fn normalize_arrow_points(world_points: &[DrawPoint], rect: DrawRect) -> Vec<DrawPoint> {
    normalize_connector_points(world_points, rect)
}

/// Calculates connector bounds from world-space points.
pub fn calculate_arrow_path_bounds_via_core(
    world_points: &[DrawPoint],
    arrow_type: ArrowType,
) -> DrawRect {
    calculate_connector_rect(world_points, arrow_type)
}

/// Recomputes geometry for arrow-core style local point edits.
pub fn resolve_core_geometry_update(
    local_points: &[DrawPoint],
    old_rect: DrawRect,
    rotation: f64,
    arrow_type: ArrowType,
) -> ConnectorGeometryUpdate {
    resolve_connector_geometry_update(local_points, old_rect, rotation, arrow_type)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn adapter_wrappers_delegate_to_connector_geometry() {
        let rect = DrawRect::new(10.0, 20.0, 110.0, 220.0);
        let world_points = vec![DrawPoint::new(10.0, 20.0), DrawPoint::new(110.0, 220.0)];

        let normalized = normalize_arrow_points(&world_points, rect);
        let resolved_world = resolve_arrow_world_points(rect, &normalized);
        let resolved_local = resolve_arrow_local_points(rect, &normalized);
        let fast_layout = compute_arrow_two_point_layout(world_points[0], world_points[1]);

        assert_eq!(resolved_world, world_points);
        assert_eq!(resolved_local, vec![DrawPoint::ZERO, DrawPoint::new(100.0, 200.0)]);
        assert_eq!(calculate_arrow_path_bounds_via_core(&world_points, ArrowType::Straight), rect);
        assert_eq!(fast_layout.rect, rect);
        assert_eq!(fast_layout.normalized_points, normalized);
    }
}
